src/main/scala/com/fulcrumgenomics/util/AssignPrimers.scala

```diff
@@ -0,0 +1,245 @@
+    |end positions of the primers for each amplicon.  The primer file must include headers, e.g:
+    |
+    |```
+    |chrom  left_start  left_end  right_start right_end
+    |chr1   1010873     1010894   1011118     1011137
+    |```
+    |
+    |Optionally, a sixth column column `id` may be given with a unique name for the amplicon.  If not given, the
+    |coordinates of the amplicon's primers will be used:
+    |  `<chrom>:<left_start>-<left_end>,<chrom>:<right_start>:<right_end>`
+    |
+    |The output will have the following tags added:
+    |- ap: the assigned primer coordinates (ex. `chr1:1010873-1010894`)
+    |- am: the mate's assigned primer coordinates (ex. `chr1:1011118-1011137`)
+    |- ip: the assigned amplicon id
+    |- im: the mate's assigned amplicon id (or `=` if the same as the assigned amplicon)
+    |
+    |The read sequence of the primer is not checked against the expected reference sequence at the primer's genomic
+    |coordinates.
+  """)
+class AssignPrimers
+(@arg(flag='i', doc="Input BAM file.")  val input: PathToBam,
+ @arg(flag='o', doc="Output BAM file.") val output: PathToBam,
+ @arg(flag='m', doc="Output metrics file.") val metrics: FilePath,
+ @arg(flag='p', doc="File with primer locations.") val primers: FilePath,
+ @arg(flag='S', doc="Match to primer locations +/- this many bases.") val slop: Int = 5,
+ @arg(flag='U', doc="True to based on the unclipped coordinates (adjust based on hard/soft clipping), otherwise the aligned bases") val unclippedCoordinates: Boolean = true,
+ @arg(doc="The SAM tag for the assigned primer coordinate.") val primerCoordinatesTag: String = AssignPrimers.PrimerCoordinateTag,
+ @arg(doc="The SAM tag for the mate's assigned primer coordinate.") val matePrimerCoordinatesTag: String = AssignPrimers.MatePrimerCoordinateTag,
+ @arg(doc="The SAM tag for the assigned amplicon identifier.") val ampliconIdentifierTag: String = AssignPrimers.AmpliconIdentifierTag,
+ @arg(doc="The SAM tag for the mate's assigned amplicon identifier.") val mateAmpliconIdentifierTag: String = AssignPrimers.MateAmpliconIdentifierTag,
+) extends FgBioTool with LazyLogging {
+
+  Io.assertReadable(input)
+  Io.assertReadable(primers)
+  Io.assertCanWriteFile(output)
+
+  override def execute(): Unit = {
+    val reader    = SamSource(input)
+    val writer    = SamWriter(output, reader.header)
+    val progress  = ProgressLogger(logger=logger, unit=250000)
+    val amplicons = Metric.read[Amplicon](path=primers)
+    val detector  = new AmpliconDetector(
+      detector             = Amplicon.detector(amplicons=amplicons.iterator),
+      slop                 = slop,
+      unclippedCoordinates = unclippedCoordinates
+    )
+
+    val metricsMap: Map[Amplicon, AssignPrimersMetric] = amplicons.map { amplicon =>
+      amplicon -> new AssignPrimersMetric(identifier=amplicon.identifier)
+    }.toMap
+
+    val labeller = new AmpliconLabeller(
```

Comment:
Seems strange to have a "labeller" be the thing updating the metrics, but not constructing the metric.  I would eithe bring the metrics updates out into the main body _or_ pass in the `Amplicons` and have it manufacture the metrics internally.